Convert MNIST PyTorch and MNIST Ray test cases to go

sutaakar opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Converting https://github.com/opendatahub-io/distributed-workloads/blob/e9ce41339e6158d6949e7fb55781647cb35f79da/tests/basictests/distributed-workloads.sh#L33-L90 into go lang integration tests.

## Description
<!--- Describe your changes in detail -->
As part of https://github.com/opendatahub-io/distributed-workloads/issues/63 this PR converts MNIST PyTorch test case into go lang.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested by executing locally against OCP.

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

anishasthana commented on tests/go.mod:
```diff
@@ -54,4 +54,6 @@ require (
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
 
+replace github.com/project-codeflare/codeflare-operator => /home/ksuta/src/github.com/project-codeflare/codeflare-operator
```
This doesn't seem right

anishasthana replied:
Ah,  i see the pr is marked wip. Sorry!

astefanutti commented on tests/integration/mcad_pytorch_test.go:
```diff
@@ -0,0 +1,123 @@
+	// Read the Notebook CR from resources and perform replacements for custom values.
+	// Currently ODH doesn't support creating Notebook CRs outside of UI, https://github.com/opendatahub-io/kubeflow/issues/126 should address this.
+	customNb := string(ReadFile(test, "resources/custom-nb-small.yaml"))
+	replacements := map[string]string{
```
Could Go template be used instead?

sutaakar replied:
Done

astefanutti commented on tests/integration/mcad_pytorch_test.go:
```diff
@@ -0,0 +1,123 @@
+	// Create Notebook CR using OC client, didn't find a Notebook go lang client for programmatic creation
+	file, err := support.CreateTempFile(test, customNb)
+	test.Expect(err).NotTo(HaveOccurred())
+	output, err := executeCommand("oc", "apply", "-n", namespace.Name, "-f", file.Name())
```
An alternative would be to use the dynamic client from the `k8s.io/client-go/dynamic` package. That would make it more portable.

sutaakar replied:
Migrated to use dynamic client

astefanutti commented on tests/integration/mcad_pytorch_test.go:
```diff
@@ -0,0 +1,123 @@
+		Should(And(HaveLen(1), ContainElement(WithTransform(support.AppWrapperState, Equal(mcadv1beta1.AppWrapperStateActive)))))
+
+	// Make sure the AppWrapper finishes and is deleted
+	test.Eventually(support.AppWrappersWithPrefix(test, namespace, "mnistjob"), support.TestTimeoutLong).
```
Could labels be used instead, so we rely on the idiomatic label selector pattern?

sutaakar replied:
Unfortunately SDK doesn't create any label or annotation in AppWrapper.
At least not in this case. 
Maybe test support should return all AppWrappers and here in test I will check that its name starts with `mnistjob`.

astefanutti commented on tests/integration/mcad_pytorch_test.go:
```diff
@@ -0,0 +1,171 @@
+			},
+		},
+	}
+	notebookPvc, err = test.Client().Core().CoreV1().PersistentVolumeClaims(namespace.Name).Create(test.Ctx(), notebookPvc, metav1.CreateOptions{})
```
nit: `notebookPvc` -> `notebookPVC`

astefanutti commented on tests/integration/mcad_pytorch_test.go:
```diff
@@ -0,0 +1,171 @@
+	notebookProps := NotebookProps{
+		IngressDomain:           getIngressDomain(test),
+		OpenShiftApiUrl:         getOpenShiftApiUrl(test),
+		KubernetesBearerToken:   getKubernetesBearerToken(test),
```
Should the Notebook be run by impersonating a typical user, rather than using what will otherwise often be cluster admin, so RBAC requirements are also covered?

sutaakar replied:
I would say it should use typical user

astefanutti replied:
Agreed. One option would be to create a ServiceAccount representing the user, bind the expected RBAC, and retrieve the token from that ServiceAccount (using the TokenRequest API for example).

There is an example of using ServiceAccount for impersonating the end-user in the CodeFlare operator tests: https://github.com/project-codeflare/codeflare-operator/blob/486c24114b1b8a451c2c9780cb128b73faf573e0/test/e2e/mnist_raycluster_sdk_test.go#L75-L140.

sutaakar replied:
Done

sutaakar commented:
@astefanutti adjusted PR:
- added a test for MCAD Ray
- replaced `oc` CLI commands with dynamic client
- minor adjustments here and there

TODO:
- change AppWrapper group - waiting for a new SDK release

There may be potential issue with MCAD Ray test for OpenShift CI as Ray head requires a lot of resources. Maybe there is a need for another node (or for a possibility to modify head node requirements). Will see once TODO is addressed.

astefanutti commented on tests/integration/mcad_test.go:
```diff
@@ -0,0 +1,362 @@
+		Should(HaveLen(0))
+}
+
+func TestMCADRay(t *testing.T) {
```
What about splitting the file a bit by moving that test case in a separate file? If yes, the shared support functions could also be moved in a separate file.

sutaakar replied:
Done.
PR was ~~completely rewritten~~ adjusted by separating support methods into dedicated files.
I have kept them so far here in ODH, if needed they can be later moved to the CFO support package.

astefanutti commented on tests/integration/mcad_test.go:
```diff
@@ -0,0 +1,362 @@
+		OpenDataHubNamespace:    GetOpenDataHubNamespace(),
+		CodeFlareImageStreamTag: getCodeFlareImageStreamTag(test),
+		JobType:                 "ray",
+		NotebookPVC:             "jupyterhub-nb-kube-3aadmin-pvc",
```
`notebookPVC.Name`?

astefanutti commented on tests/integration/mcad_test.go:
```diff
@@ -0,0 +1,362 @@
+		OpenDataHubNamespace:    GetOpenDataHubNamespace(),
+		CodeFlareImageStreamTag: getCodeFlareImageStreamTag(test),
+		JobType:                 "mcad",
+		NotebookPVC:             "jupyterhub-nb-kube-3aadmin-pvc",
```
`notebookPVC.Name`?

astefanutti commented on tests/integration/mcad_test.go:
```diff
@@ -0,0 +1,362 @@
+	cluster, err := test.Client().Dynamic().Resource(ingressConfigResource).Get(test.Ctx(), "cluster", metav1.GetOptions{})
+	test.Expect(err).NotTo(HaveOccurred())
+
+	ingressDomain := cluster.UnstructuredContent()["spec"].(map[string]interface{})["domain"].(string)
```
I'd suggest to use `unstructured.NestedString`.

astefanutti commented on tests/integration/mcad_test.go:
```diff
@@ -0,0 +1,362 @@
+	cluster, err := test.Client().Dynamic().Resource(infrastructureConfigResource).Get(test.Ctx(), "cluster", metav1.GetOptions{})
+	test.Expect(err).NotTo(HaveOccurred())
+
+	openShiftApiUrl := cluster.UnstructuredContent()["status"].(map[string]interface{})["apiServerURL"].(string)
```
I'd suggest to use `unstructured.NestedString`.

astefanutti reviewed (commented):
Awesome. I've left a couple of small comments. Otherwise I'll LTGM it once ready.

astefanutti commented on tests/integration/support/net.go:
```diff
@@ -0,0 +1,60 @@
+	cluster, err := test.Client().Dynamic().Resource(ingressConfigResource).Get(test.Ctx(), "cluster", metav1.GetOptions{})
+	test.Expect(err).NotTo(gomega.HaveOccurred())
+
+	spec, found, err := unstructured.NestedMap(cluster.UnstructuredContent(), "spec")
+	test.Expect(err).NotTo(gomega.HaveOccurred())
+	test.Expect(found).To(gomega.BeTrue())
+
+	ingressDomain, found, err := unstructured.NestedString(spec, "domain")
+	test.Expect(err).NotTo(gomega.HaveOccurred())
+	test.Expect(found).To(gomega.BeTrue())
```
```suggestion
	ingressDomain, found, err := unstructured.NestedString(cluster.UnstructuredContent(), "spec", "domain")
	test.Expect(err).NotTo(gomega.HaveOccurred())
	test.Expect(found).To(gomega.BeTrue())
```

astefanutti commented on tests/integration/support/net.go:
```diff
@@ -0,0 +1,60 @@
+	cluster, err := test.Client().Dynamic().Resource(infrastructureConfigResource).Get(test.Ctx(), "cluster", metav1.GetOptions{})
+	test.Expect(err).NotTo(gomega.HaveOccurred())
+
+	status, found, err := unstructured.NestedMap(cluster.UnstructuredContent(), "status")
+	test.Expect(err).NotTo(gomega.HaveOccurred())
+	test.Expect(found).To(gomega.BeTrue())
+
+	openShiftApiUrl, found, err := unstructured.NestedString(status, "apiServerURL")
+	test.Expect(err).NotTo(gomega.HaveOccurred())
+	test.Expect(found).To(gomega.BeTrue())
```
```suggestion
	openShiftApiUrl, found, err := unstructured.NestedString(cluster.UnstructuredContent(), "status", "apiServerURL")
	test.Expect(err).NotTo(gomega.HaveOccurred())
	test.Expect(found).To(gomega.BeTrue())
```

astefanutti commented on tests/integration/support/net.go:
```diff
@@ -0,0 +1,60 @@
+var ingressConfigResource = schema.GroupVersionResource{Group: "config.openshift.io", Version: "v1", Resource: "ingresses"}
+var infrastructureConfigResource = schema.GroupVersionResource{Group: "config.openshift.io", Version: "v1", Resource: "infrastructures"}
+
+func GetIngressDomain(test cfosupport.Test) string {
```
Maybe `test.T().Helper()`?

astefanutti commented on tests/integration/support/net.go:
```diff
@@ -0,0 +1,60 @@
+	return ingressDomain
+}
+
+func GetOpenShiftApiUrl(test cfosupport.Test) string {
```
Maybe `test.T().Helper()`?

astefanutti commented on tests/integration/support/odh.go:
```diff
@@ -0,0 +1,30 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func GetODHCodeFlareImageStreamTag(test cfosupport.Test) string {
```
Maybe `test.T().Helper()`?

astefanutti commented on tests/integration/support/rbac.go:
```diff
@@ -0,0 +1,90 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func CreateTestRBAC(test cfosupport.Test, namespace *corev1.Namespace, policyRules []rbacv1.PolicyRule) (token string) {
```
Maybe `test.T().Helper()`?

sutaakar commented:
Changes applied.

astefanutti commented:
/lgtm

astefanutti commented:
/lgtm

sutaakar commented:
Failing because the ODH core KfDef/DSC is not installed in OpenShift CI.
Should be fixed either by https://github.com/openshift/release/pull/43247 or once I migrate the ODH DW to use DSC.

sutaakar commented:
/retest

sutaakar commented:
/retest

sutaakar commented:
/retest

sutaakar commented:
/retest

sutaakar commented:
`TestMnistPyTorchMCAD` passed, `TestMCADRay` failed, supposedly on lack of resources to spin up the Ray head node

sutaakar commented:
Skipping unstable/currently failing tests.

sutaakar commented:
/retest

sutaakar commented:
@astefanutti Can you please take a final look/provide approval?

astefanutti commented:
/lgtm

astefanutti commented:
/approve

sutaakar commented:
/retest
